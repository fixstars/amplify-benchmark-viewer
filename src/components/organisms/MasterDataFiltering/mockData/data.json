[
  { "name": "Tsp" },
  { "name": "Qap" },
  { "name": "MaxCut" },
  { "name": "Sudoku" },
  { "name": "Cvrp" }
]
