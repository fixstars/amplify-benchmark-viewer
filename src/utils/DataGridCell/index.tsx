/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { LinkButton } from 'components/atoms/LinkButton'

const clientColors = [
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
]

const instanceColors = [
  '#ff66ff',
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#718de6',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
  '#9a916d',
  '#406a0f',
  '#004b31',
  '#18729f',
  '#9997c3',
  '#9cf0ec',
  '#ce843c',
  '#2d326e',
  '#76ed60',
  '#d7f7e8',
  '#bb15d7',
  '#87a7d8',
  '#94053a',
  '#e5c8a6',
  '#e2f2c0',
]

export interface ProblemInstances {
  readonly [key: string]: ReadonlyArray<string> | undefined
}

const renderClientsCell = (key: string, clients: ReadonlyArray<string>) => {
  const getBackgroundColor = (client: string): string => {
    const cache = JSON.parse(localStorage.getItem('clientColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('clientColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[client]

    if (backgroundColor === undefined) {
      const length = clientColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = clientColors[cacheIndex]
      cache[client] = backgroundColor
      localStorage.setItem('clientColorCache', JSON.stringify(cache))
      localStorage.setItem('clientColorIndexCache', JSON.stringify(cacheIndex))
      return backgroundColor
    }

    return backgroundColor
  }

  return clients.map((client) => (
    <div key={`${key}_${client}`} style={{ marginRight: 10 }}>
      <LinkButton
        label={client}
        link={`/clients/${client}`}
        backgroundColor={getBackgroundColor(client)}
      />
    </div>
  ))
}

const renderProblemInstancesCell = (
  key: string,
  problemInstances: ProblemInstances,
) => {
  const getBackgroundColor = (instance: string): string => {
    const cache = JSON.parse(localStorage.getItem('instanceColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('instanceColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[instance]

    if (backgroundColor === undefined) {
      const length = instanceColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = instanceColors[cacheIndex]
      cache[instance] = backgroundColor
      localStorage.setItem('instanceColorCache', JSON.stringify(cache))
      localStorage.setItem(
        'instanceColorIndexCache',
        JSON.stringify(cacheIndex),
      )
      return backgroundColor
    }

    return backgroundColor
  }
  const pairs = Object.entries(problemInstances)
  pairs.sort(function (p1, p2) {
    const p1Key = p1[0]
    const p2Key = p2[0]
    if (p1Key < p2Key) {
      return -1
    }
    if (p1Key > p2Key) {
      return 1
    }
    return 0
  })
  problemInstances = Object.fromEntries(pairs)
  return Object.keys(problemInstances).map((className) => (
    <div
      key={`${key}_${className}`}
      style={{ margin: '5px 0 5px 0', display: 'flex' }}
    >
      <span style={{ marginRight: 10 }}>{className}:</span>
      <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        {problemInstances[className]?.map((instance) => (
          <div
            key={`${key}_${className}_${instance}`}
            style={{ marginRight: 10, marginBottom: 10 }}
          >
            <LinkButton
              label={instance}
              link={`/classes/${className}/instances/${instance}`}
              backgroundColor={getBackgroundColor(instance)}
            />
          </div>
        ))}
      </div>
    </div>
  ))
}

export { renderClientsCell, renderProblemInstancesCell }
