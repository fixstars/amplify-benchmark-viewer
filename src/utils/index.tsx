export * from './DataGridCell'
export * from './statistics'
