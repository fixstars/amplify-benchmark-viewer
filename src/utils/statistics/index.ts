function quartile(data: readonly number[], q: number) {
  const sortedData = data.slice().sort((a, b) => a - b)
  const pos = Math.floor((sortedData.length - 1) * q)
  return sortedData[pos]
}
function quartile0(data: readonly number[]) {
  return quartile(data, 0)
}
function quartile25(data: readonly number[]) {
  return quartile(data, 0.25)
}

function quartile50(data: readonly number[]) {
  return quartile(data, 0.5)
}

function quartile75(data: readonly number[]) {
  return quartile(data, 0.75)
}

function quartile100(data: readonly number[]) {
  return quartile(data, 1)
}

export { quartile0, quartile25, quartile50, quartile75, quartile100 }
