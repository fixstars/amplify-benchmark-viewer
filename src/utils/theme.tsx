/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { createTheme } from '@mui/material/styles'

export const theme = createTheme({
  palette: {
    primary: {
      main: '#111',
    },
  },
  typography: {
    h2: {
      fontSize: '1.5rem',
    },
  },
  components: {
    MuiTooltip: {
      styleOverrides: {
        tooltip: {
          backgroundColor: '#111',
          color: '#f4f1f4',
          font: '10px',
        },
      },
    },
  },
})
