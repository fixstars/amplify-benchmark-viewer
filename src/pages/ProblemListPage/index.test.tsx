import { render } from '@testing-library/react'
import type { ReportData } from '@types'
import { AxiosError } from 'axios'
import { createMemoryHistory } from 'history'
import { BrowserRouter, Router } from 'react-router-dom'
import { RecoilRoot } from 'recoil'
import { mockLayout, mockProblemList } from 'utils/test'
import mockData from 'utils/test/mocks/data/mockReportData.json'

import { ProblemListPage } from '.'

let mockResponse: {
  readonly data: ReportData | undefined
  readonly isLoading: boolean
  readonly isFetching: boolean
  readonly isError: boolean
  readonly error: AxiosError | null
} = {
  data: undefined,
  isLoading: false,
  isFetching: false,
  isError: false,
  error: null,
}
jest.mock('api', () => ({
  useGetReportData: () => mockResponse,
}))

describe('<ProblemListPage />', () => {
  beforeEach(() => {
    mockResponse = {
      data: undefined,
      isLoading: false,
      isFetching: false,
      isError: false,
      error: null,
    }
  })

  it('rendered well with data', () => {
    mockResponse = {
      ...mockResponse,
      data: mockData,
    }

    const { container } = render(
      <RecoilRoot>
        <BrowserRouter>
          <ProblemListPage />
        </BrowserRouter>
      </RecoilRoot>,
    )

    expect(mockProblemList.mock.calls.length).toBe(1)
    expect(mockProblemList.mock.calls[0][0].data).toEqual([
      { class: 'Tsp', instance: 'burma14', clients: ['GurobiClient'] },
      { class: 'Tsp', instance: 'pr124', clients: ['FixstarsClient'] },
      { class: 'Tsp', instance: 'pr136', clients: ['FixstarsClient'] },
      { class: 'Tsp', instance: 'pr144', clients: ['FixstarsClient'] },
      { class: 'Tsp', instance: 'ulysses16', clients: ['GurobiClient'] },
    ])

    expect(container).toMatchSnapshot()
  })

  it('rendered well without data', () => {
    mockResponse = { ...mockResponse, data: undefined }

    const { container } = render(
      <RecoilRoot>
        <BrowserRouter>
          <ProblemListPage />
        </BrowserRouter>
      </RecoilRoot>,
    )

    expect(mockLayout.mock.calls.length).toBe(1)
    const layout = mockLayout.mock.calls[0][0].children
    expect(layout.type.name).toBe('Loading')

    expect(mockProblemList.mock.calls.length).toBe(0)

    expect(container).toMatchSnapshot()
  })

  it('isLoading', () => {
    mockResponse = {
      ...mockResponse,
      data: mockData,
      isLoading: true,
    }

    render(
      <RecoilRoot>
        <BrowserRouter>
          <ProblemListPage />
        </BrowserRouter>
      </RecoilRoot>,
    )

    expect(mockLayout.mock.calls.length).toBe(2)
    const layout = mockLayout.mock.calls[1][0].children
    expect(layout.type.name).toBe('Loading')

    expect(mockProblemList.mock.calls.length).toBe(0)
  })

  it('isFetching', () => {
    mockResponse = {
      ...mockResponse,
      data: mockData,
      isFetching: true,
    }

    render(
      <RecoilRoot>
        <BrowserRouter>
          <ProblemListPage />
        </BrowserRouter>
      </RecoilRoot>,
    )

    expect(mockLayout.mock.calls.length).toBe(2)
    const layout = mockLayout.mock.calls[1][0].children
    expect(layout.type.name).toBe('Loading')

    expect(mockProblemList.mock.calls.length).toBe(0)
  })

  it('Redirect to upload page if no data', () => {
    const mockPush = jest.fn()
    const history = createMemoryHistory({ initialEntries: ['/classes'] })
    history.push = mockPush

    mockResponse = {
      ...mockResponse,
      data: undefined,
      isError: true,
      error: new AxiosError('Some network error'),
    }

    render(
      <RecoilRoot>
        <Router location={history.location} navigator={history}>
          <ProblemListPage />
        </Router>
      </RecoilRoot>,
    )

    expect(mockLayout.mock.calls.length).toBe(1)
    const layout = mockLayout.mock.calls[0][0].children
    expect(layout.type.name).toBe('Loading')
    expect(mockProblemList.mock.calls.length).toBe(0)

    // redirect
    expect(mockPush.mock.calls[0][0].pathname).toBe('/upload')
  })
})
