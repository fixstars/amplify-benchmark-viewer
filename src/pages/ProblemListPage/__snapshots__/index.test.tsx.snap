// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProblemListPage /> rendered well with data 1`] = `
<div>
  <div
    class="MuiBox-root css-thz6x0"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1hc7nu0-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1r339ym-MuiToolbar-root"
      >
        <div
          class="css-yp9swi"
        >
          <a
            class="css-94l1l1"
            href="/"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-19sbe58-MuiTypography-root"
            >
              Annealing Benchmark
            </h6>
          </a>
        </div>
        <nav
          class="css-ov1ktg"
        >
          <a
            class="css-94l1l1"
            href="/"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-4hrd4f-MuiTypography-root"
            >
              Problems
            </p>
          </a>
          <a
            class="css-94l1l1"
            href="/clients"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
            >
              Clients
            </p>
          </a>
          <a
            class="css-94l1l1"
            href="/labels"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
            >
              Label
            </p>
          </a>
        </nav>
      </div>
    </header>
    <div
      class="MuiBox-root css-1cf7osc"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      />
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-15h1z56-MuiGrid-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h2 css-1asg1ln-MuiTypography-root"
          >
            Problem List
          </h2>
        </div>
      </div>
      <div
        class="MuiBox-root css-13pmxen"
      >
        <div
          aria-colcount="3"
          aria-multiselectable="false"
          aria-rowcount="6"
          class="MuiDataGrid-root MuiDataGrid-root--densityStandard MuiDataGrid-withBorderColor css-11g4t4x-MuiDataGrid-root"
          role="grid"
        >
          <div
            class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
          >
            <div
              class="MuiDataGrid-columnHeaders MuiDataGrid-withBorderColor css-1iyq7zh-MuiDataGrid-columnHeaders"
              style="min-height: 56px; max-height: 56px; line-height: 56px;"
            >
              <div
                class="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-gl260s-MuiDataGrid-columnHeadersInner"
                role="rowgroup"
                style="transform: translate3d(0px, 0px, 0px);"
              >
                <div
                  aria-rowindex="1"
                  class="css-yrdy0g-MuiDataGrid-columnHeaderRow"
                  role="row"
                >
                  <div
                    aria-colindex="1"
                    aria-label="Class"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader MuiDataGrid-withBorderColor"
                    data-field="class"
                    role="columnheader"
                    style="height: 56px; width: 180px; min-width: 180px; max-width: 180px;"
                    tabindex="0"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Class
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    aria-colindex="2"
                    aria-label="Instance"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader MuiDataGrid-withBorderColor"
                    data-field="instance"
                    role="columnheader"
                    style="height: 56px; width: 180px; min-width: 180px; max-width: 180px;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Instance
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    aria-colindex="3"
                    aria-label="Clients"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader MuiDataGrid-withBorderColor"
                    data-field="clients"
                    role="columnheader"
                    style="height: 56px; width: 0px; min-width: 0; max-width: 0;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Clients
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiDataGrid-virtualScroller css-axafay-MuiDataGrid-virtualScroller"
              style="overflow-x: hidden;"
            >
              <div
                class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
                style="width: auto; height: 260px; min-height: auto;"
              >
                <div
                  class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
                  style="transform: translate3d(0px, 0px, 0px);"
                >
                  <div
                    aria-rowindex="2"
                    aria-selected="false"
                    class="MuiDataGrid-row"
                    data-id="Tsp_a_instance"
                    data-rowindex="0"
                    role="row"
                    style="max-height: 52px; min-height: 52px;"
                  >
                    <div
                      aria-colindex="1"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="0"
                      data-field="class"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        title="Tsp"
                      >
                        Tsp
                      </div>
                    </div>
                    <div
                      aria-colindex="2"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="1"
                      data-field="instance"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <a
                        href="/classes/Tsp/instances/a_instance"
                      >
                        a_instance
                      </a>
                    </div>
                    <div
                      aria-colindex="3"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="2"
                      data-field="clients"
                      role="cell"
                      style="min-width: 0; max-width: 0; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        style="margin-right: 10px;"
                      >
                        <a
                          class="css-1o7gueb"
                          href="/clients/a_client"
                        >
                          a_client
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    aria-rowindex="3"
                    aria-selected="false"
                    class="MuiDataGrid-row"
                    data-id="Tsp_b_instance"
                    data-rowindex="1"
                    role="row"
                    style="max-height: 52px; min-height: 52px;"
                  >
                    <div
                      aria-colindex="1"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="0"
                      data-field="class"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        title="Tsp"
                      >
                        Tsp
                      </div>
                    </div>
                    <div
                      aria-colindex="2"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="1"
                      data-field="instance"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <a
                        href="/classes/Tsp/instances/b_instance"
                      >
                        b_instance
                      </a>
                    </div>
                    <div
                      aria-colindex="3"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="2"
                      data-field="clients"
                      role="cell"
                      style="min-width: 0; max-width: 0; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        style="margin-right: 10px;"
                      >
                        <a
                          class="css-1o7gueb"
                          href="/clients/a_client"
                        >
                          a_client
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    aria-rowindex="4"
                    aria-selected="false"
                    class="MuiDataGrid-row"
                    data-id="Tsp_h_instance"
                    data-rowindex="2"
                    role="row"
                    style="max-height: 52px; min-height: 52px;"
                  >
                    <div
                      aria-colindex="1"
                      aria-colspan="1"
                      class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="0"
                      data-field="class"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        class="MuiDataGrid-cellContent"
                        title="Tsp"
                      >
                        Tsp
                      </div>
                    </div>
                    <div
                      aria-colindex="2"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="1"
                      data-field="instance"
                      role="cell"
                      style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <a
                        href="/classes/Tsp/instances/h_instance"
                      >
                        h_instance
                      </a>
                    </div>
                    <div
                      aria-colindex="3"
                      aria-colspan="1"
                      class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                      data-colindex="2"
                      data-field="clients"
                      role="cell"
                      style="min-width: 0; max-width: 0; min-height: 52px; max-height: 52px;"
                      tabindex="-1"
                    >
                      <div
                        style="margin-right: 10px;"
                      >
                        <a
                          class="css-gqb4bp"
                          href="/clients/f_client"
                        >
                          f_client
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ProblemListPage /> rendered well without data 1`] = `
<div>
  <div
    class="MuiBox-root css-thz6x0"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1hc7nu0-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1r339ym-MuiToolbar-root"
      >
        <div
          class="css-yp9swi"
        >
          <a
            class="css-94l1l1"
            href="/"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-19sbe58-MuiTypography-root"
            >
              Annealing Benchmark
            </h6>
          </a>
        </div>
        <nav
          class="css-ov1ktg"
        >
          <a
            class="css-94l1l1"
            href="/"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-4hrd4f-MuiTypography-root"
            >
              Problems
            </p>
          </a>
          <a
            class="css-94l1l1"
            href="/clients"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
            >
              Clients
            </p>
          </a>
          <a
            class="css-94l1l1"
            href="/labels"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
            >
              Label
            </p>
          </a>
        </nav>
      </div>
    </header>
    <div
      class="MuiBox-root css-1cf7osc"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      />
      <div
        class="MuiBox-root css-x7m8p7"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
